src/SurfaceSchemes/surface_conditions.jl

```diff
@@ -0,0 +1,351 @@
+# TODO: Remove this type piracy.
+Base.Broadcast.BroadcastStyle(
+    ::Base.Broadcast.Style{Tuple},
+    ds::DataLayouts.DataStyle,
+) = ds
+
+"""
+    update_surface_conditions!(Y, p, t)
+
+Updates the value of `p.sfc_conditions` based on the current state `Y` and time
+`t`. This function will usually be called automatically, but it needs to be
+called manually if `surface_updater` is set to `ManualUpdates`.
+"""
+function update_surface_conditions!(Y, p, t)
+    C1 = Geometry.Covariant1Vector
+    C2 = Geometry.Covariant2Vector
+    (; ᶜts, sfc_conditions, sfc_state, params, atmos) = p
+    sfc_local_geometry = Fields.field_values(
+        Fields.level(Fields.local_geometry_field(Y.f), Fields.half),
+    )
+    int_ts = Fields.field_values(Fields.level(ᶜts, 1))
+    int_uₕ = Fields.field_values(Fields.level(Y.c.uₕ, 1))
+    int_z = Fields.field_values(Fields.level(Fields.coordinate_field(Y.c).z, 1))
+    sfc_conditions_values = Fields.field_values(sfc_conditions)
+
+    #=
+    sfc_state_broadcast =
+        if sfc_state isa Function
+            Base.Broadcast.Broadcasted(sfc_state, sfc_local_geometry.coordinates, t)
+        elseif sfc_state isa Fields.Field
+            Fields.field_values(sfc_state)
+        elseif sfc_state isa SurfaceState
+            sfc_state
+        end
+    =#
+    sfc_state_broadcast = if isa(sfc_state, Function)
+        (sfc_state(sfc_local_geometry.coordinates, int_z, t),)
+    elseif isa(sfc_state, Fields.Field)
+        Fields.field_values(sfc_state)
+    elseif isa(sfc_state, SurfaceState)
+        (sfc_state,)
+    end
+    @. sfc_conditions_values = surface_conditions(
+        sfc_state_broadcast,
+        sfc_local_geometry,
+        int_ts,
+        int_uₕ.components.data.:1 /
+        unit_basis_vector_data(C1, sfc_local_geometry),
+        int_uₕ.components.data.:2 /
+        unit_basis_vector_data(C2, sfc_local_geometry),
```

Comment:
I talked with Dennis offline. I'm not sure if this is correct. We can revisit it.